### PR TITLE
CATROID-831 Add a search tool to scripts view

### PR DESCRIPTION
**Pull request by devsebastian**
Ticket: https://jira.catrob.at/browse/CATROID-831

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

**Comment by devsebastian** [comment]
I've used a different approach than what is discussed in the ticket and it work's perfectly fine at the moment, let me know if the current approach needs to be improved or changed! Happy to update the code accordingly

**Review by Walcho1125 — CHANGES_REQUESTED** [review]
Well done! 
Please resolve the conflicts that occur. Additionally please remove the find option from the Category view when adding a brick, since the search option should only be available in the script view.
Best Regards!

**Review by Walcho1125 — APPROVED** [review]
Well done, looks good to me!

File: catroid/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -2113,9 +2114,11 @@ needs read and write access to it. You can always change permissions through you
     data is turned on.</string>
     <string name="edit">Edit</string>
     <string name="data_value">Value</string>
-    <string name="category_recently_used">Recently used</string>
 
     <!-- Information -->
     <string name="no_variable_selected">No variable selected</string>
+    <string name="find">Find</string>
+    <string name="find_next">Find Next %d/%d</string>
```
Comment (wslany) [inline catroid/src/main/res/values/strings.xml:2121]:
```suggestion
    <string name="find_next">Find next %d/%d</string>
```

**Review by wslany — CHANGES_REQUESTED** [review]
Thanks! Great that it works with the translated formula content!!

Currently the search only works in the current object's scripts. Instead, it should search through all objects in all scenes, and thus also jump to another object and place in scripts where the phrase is found.

Also please search through pull-down type (spinner) parameter fields such as names of scenes, objects, variables, lists, broadcast messages, instruments, drums, looks, backgrounds, or sounds.

Also, if it is easy, please make the search case insensitive (unicode character encoding based if supported, see https://developer.android.com/reference/java/util/regex/Pattern#UNICODE_CASE --- for regular expressions it is possible to use the (?u) prefix).

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ScriptFragment.java [hunk 2]
```diff
@@ -243,11 +256,64 @@ private void resetActionModeParameters() {
...
+
+		scriptFinder.setOnResultFoundListener((sceneIndex, spriteIndex, brickIndex, totalResults) -> {
+
+			Project currentProject = ProjectManager.getInstance().getCurrentProject();
+			Scene currentScene = currentProject.getSceneList().get(sceneIndex);
+			Sprite currentSprite = currentScene.getSpriteList().get(spriteIndex);
+
+//            ((BaseActivity) getActivity()).getSupportActionBar().setTitle(String.format("%s - %s",
+//                    getString(R.string.search),
+//                    currentSprite.getName()));
```
Comment (Walcho1125) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ScriptFragment.java:270]: please remove that out commented code

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ScriptFragment.java [hunk 3]
```diff
@@ -452,6 +522,7 @@ public void onSelectionChanged(int selectedItemCnt) {
 				break;
 			case NONE:
 				throw new IllegalStateException("ActionModeType not set Correctly");
+			case FIND:
```
Comment (Walcho1125) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ScriptFragment.java:525]: break is missing

**Review by Walcho1125 — CHANGES_REQUESTED** [review]
Please take a look at my small remarks, the functionality works out good!

**Review by Walcho1125 — DISMISSED** [review]
Please resolve the merge conflicts!

**Review by Walcho1125 — APPROVED** [review]
well done! looks good to me now

**Review by wslany — CHANGES_REQUESTED** [review]
Hi Dev, after trying the interface, I found the checkmark button not intuitive --- I kept tapping on it to start the search. How about the following: Instead of the checkmark, move the search field and the magnification icon into the top bar. The "Find keyword" should be replaced by a larger text field. Like now, the v and ^ buttons replace the magnification button once more than one result has been found to navigate among the results. What do you think? I believe it is still material design conform, since I have seen similar ways to place the search field in several other apps as well. I.e., 
image

**Review by wslany — APPROVED** [review]

**Comment by wslany** [comment]
Thanks!!!

**Comment by mstoeg** [comment]
I think this ticket introduced a failing testcase `testForToast()` in the `SystemOutTest.java` file. @devsebastian could you please have a look at it and fix it? It is because of the Toast messages you added in the `ScriptFinder`.

**Comment by devsebastian** [comment]
> I think this ticket introduced a failing testcase `testForToast()` in the `SystemOutTest.java` file. @devsebastian could you please have a look at it and fix it? It is because of the Toast messages you added in the `ScriptFinder`.

On it 👍

**Pull request merged**